extension/observer/ecsobserver/error.go

```diff
@@ -0,0 +1,148 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecsobserver
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// error.go defines common error interfaces and util methods for generating reports
+// for log and metrics that can be used for debugging.
+
+type errWithAttributes interface {
+	// message does not include attributes like task arn etc.
+	// and expect the caller extract them using getters.
+	message() string
+	// zapFields will be logged as json attribute and allows searching and filter backend like cloudwatch.
+	// For example { $.ErrScope == "Target" } list all the error whose scope is a (scrape) target.
+	zapFields() []zap.Field
+}
+
+type taskError interface {
+	errWithAttributes
+	setTask(t *Task)
+	getTask() *Task
+}
+
+// baseTaskError can be embedded into error that happens at task level or below.
+type baseTaskError struct {
+	t *Task
+}
+
+func (be *baseTaskError) setTask(t *Task) {
```

Comment:
Feels like there might be a better way to do this without the gets and setters. If you have a lower level error but then need to annotate it with extra information like task can you use error wrapping to wrap the lower level error with a higher level error that has extra fields.

see https://blog.golang.org/go1.13-errors

Reply:
I find it a bit annoying when I am writing setter methods as well ...  

I think I can do something like

```go
type taskError struct {
    task *Task
    cause error // for unwrap
}

type containerError struct {
     task *Task
     container *Container
     cause error // could use a taskError as cause to skip saving the task, but it is not very accurate because the container caused the error, the task is just there to provide more context.
}

type targetError struct {
     task *Task
     container *Container
     target MatchTarget // just stack all the info when the error cause goes deeper...
     cause error
}
```

Another way is instead of defining types, use the pattern like `context`  package to attach attributes.

```go
type errorWithAttr interface {
      GetAttr(key string) (val interface{}, ok bool) 
}

type errorAttrWrapper struct {
      inner error
      key string
      val interface{}
}

func (ew *errorAttrWrapper) GetAttr(key string) (interface{}, bool) {
      if key == ew.key {
             return ew.val, true
      }
      ew2, ok := ew.inner.(errorWithAttr)
      if !ok {
             return nil, false
      }
      retuirn ew2.GetAttr(key)
}

func WithErrAttr(err error, key string, value interface{}) error {
     // ...
}

// And use it like
if err := exportContainer(); err != nil {
     multierr.AppendInto(merr, WithErrAttr(err, "Container", container))
}

// Reporting error
for _, err := range errors {
     v, ok := err.GetAttr("Task")
     task := v.(*Task)
     logFields = append(logFields, zap.String("TaskArn", task.Arn)
}
```

The second one is not that type safe but avoid boilerplate code for defining types

Reply:
I took the second approach, added a `internal/errctx` package, I checked https://github.com/ansel1/merry and  https://github.com/mvndaai/ctxerr and feel their interface is has too much complex for my use case.  

Use `fmt.Errorf("%w")` alone can't provide structured way for extracting info. Defining error struct end up defining several structs with similar fields (though embed would help a bit)  as shown in previous comment.